Support for JSON Web Key Sets (#505)
cakephp/authentication: docs/en/authenticators.rst

        $this->viewBuilder()->setOption('serialize', 'json');
    }

> Next to using a locally available public key making use of a JWKS fetched

Reviewer: ```suggestion
Next to use a locally available public key making use of a JWKS fetched
```

from an external JWKS endpoint is supported as well::

    // Application.php
    public function getAuthenticationService(ServerRequestInterface $request): AuthenticationServiceInterface
    {
        $service = new AuthenticationService();
        // ...
        $service->loadIdentifier('Authentication.JwtSubject');
        $service->loadAuthenticator('Authentication.Jwt', [

>             'jsonWebKeySet' => ['keys' => []],

Reviewer: ```suggestion
           // For example, the JSON decoded response of this endpoint: https://www.gstatic.com/iap/verify/public_key-jwk
            'jsonWebKeySet' => ['keys' => [...]],
```

    }

Using a JWKS fetched from an external JWKS endpoint is supported as well::

    // Application.php
    public function getAuthenticationService(ServerRequestInterface $request): AuthenticationServiceInterface
    {
        $service = new AuthenticationService();
        // ...
        $service->loadIdentifier('Authentication.JwtSubject');

        $jwksUrl = 'https://appleid.apple.com/auth/keys';

        // Set of keys. The "keys" key is required. Additionally keys require a "alg" key.
        // Add it manually to your JWK array if it doesn't already exist.
        $jsonWebKeySet = Cache::remember('jwks-' . md5($jwksUrl), function () use ($jwksUrl) {
            $http = new Client();
            $response = $http->get($jwksUrl);
            return $response->getJson();
        });

        $service->loadAuthenticator('Authentication.Jwt', [
            'jsonWebKeySet' => $jsonWebKeySet,
            'returnPayload' => false
        ]);
    }

The JWKS resource will return the same set of keys most of the time.
Applications should cache these resources, but they also need to be
prepared to handle signing key rotations.

.. warning::

>     An application would need to decide on a cache lifetime that strikes a good
>     balance between performance and security, particularly for the emergency
>     case where a private key is compromised.

Reviewer 2: ```suggestion
    Applications need pick a cache lifetime that balances performance and security.
    This is particularly important in situations where a private key is compromised.
```
I'm only reviewing the documentation changes.

Author: @othercorey - isn't it "Applications need **to** pick"

Reviewer 2: Yes, missing `to`.

        $this->viewBuilder()->setOption('serialize', 'json');
    }

Using a JWKS fetched from an external JWKS endpoint is supported as well::

    // Application.php
    public function getAuthenticationService(ServerRequestInterface $request): AuthenticationServiceInterface
    {
        $service = new AuthenticationService();
        // ...
        $service->loadIdentifier('Authentication.JwtSubject');

        $jwksUrl = 'https://appleid.apple.com/auth/keys';

        // Set of keys. The "keys" key is required. Additionally keys require a "alg" key.
        // Add it manually to your JWK array if it doesn't already exist.
        $jsonWebKeySet = Cache::remember('jwks-' . md5($jwksUrl), function () use ($jwksUrl) {
            $http = new Client();
            $response = $http->get($jwksUrl);
            return $response->getJson();
        });

        $service->loadAuthenticator('Authentication.Jwt', [
            'jsonWebKeySet' => $jsonWebKeySet,
            'returnPayload' => false
        ]);
    }

> .. note::

Reviewer 3: Does this need to be a callout? Having two callouts beside each other is something we they to avoid.

        $this->viewBuilder()->setOption('serialize', 'json');
    }

> Next to use a locally available public key making use of a JWKS fetched
> from an external JWKS endpoint is supported as well::

Reviewer 2: ```suggestion
Also supported is using a locally available public key with a JWKS fetched
from an external JWKS endpoint::
```
This might be simple to read, if not, needs a `,` after `Next`.

Author: This is going into the wrong direction - this is meant to be an alternative to the way described above.

To make it simple - this should be enough, right?

> Making use of a JWKS fetched from an external JWKS endpoint is supported as well::

    // Application.php
    public function getAuthenticationService(ServerRequestInterface $request): AuthenticationServiceInterface
    {
        $service = new AuthenticationService();
        // ...
        $service->loadIdentifier('Authentication.JwtSubject');

        $jwksUrl = 'https://appleid.apple.com/auth/keys';

>         // Set of keys. The "keys" key is required. Additionally keys require a "alg" key.

Author: The requirement for the "alg" key will be added in firebase/php-jwt:

Adding it here directly allows to make use of it for automatic creation of Key objects (see JwtAuthenticator)

https://github.com/firebase/php-jwt/pull/376/files#diff-374f5998b3c572d86be0e79432aac3de362c79e8fb146b9ce422dc2388cdc5daR53

        // Add it manually to your JWK array if it doesn't already exist.
        $jsonWebKeySet = Cache::remember('jwks-' . md5($jwksUrl), function () use ($jwksUrl) {
            $http = new Client();
            $response = $http->get($jwksUrl);
            return $response->getJson();
        });

        $service->loadAuthenticator('Authentication.Jwt', [
            'jsonWebKeySet' => $jsonWebKeySet,
            'returnPayload' => false
        ]);
    }

.. note::

    The JWKS resource will return the same set of keys most of the time.
    Applications should cache these resources, but they also need to be

>     prepared to handle signing key rotations.

Author: Added recommendations from here:

https://community.auth0.com/t/caching-jwks-signing-key/17654/2

.. warning::

    An application would need to decide on a cache lifetime that strikes a good
    balance between performance and security, particularly for the emergency

>     case where a private key is compromised.

Author: Added warnings from here:

https://community.auth0.com/t/caching-jwks-signing-key/17654/2
